### PR TITLE
searchsorted, repeat broken off from #24024

### PR DESCRIPTION
**Pull request by jbrockmendel**
Along with fixes for check_compatible_with discussed in #24024, and additional tests for `searchsorted` and `_check_compatible_with`

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -80,6 +81,79 @@ def _get_attributes_dict(self):
...
+
+        Examples
+        --------
+        >>> self._unbox_scalar(Timedelta('10s'))  # DOCTEST: +SKIP
+        10000000000
+        """
+        raise AbstractMethodError(self)
+
+    def _check_compatible_with(self, other):
+        # TODO: choose a type for other
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: I think the type is now `# type: (Union[Period, Timestamp, Timedelta, NaTType]) -> None`

Maybe note in the docstring that NaT is considered OK.

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -468,6 +542,67 @@ def _values_for_factorize(self):
     def _from_factorized(cls, values, original):
         return cls(values, dtype=original.dtype)
 
+    def _values_for_argsort(self):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: not a big  deal, but did you intend to include values_for_argsort? Maybe update the title?

Though I suspect this isn't being tested yet.

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -468,6 +542,67 @@ def _values_for_factorize(self):
...
+        return self._data
+
+    # ------------------------------------------------------------------
+    # Additional array methods
+    #  These are not part of the EA API, but we implement them because
+    #  pandas assumes they're there.
+
+    def searchsorted(self, value, side='left', sorter=None):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: Hold off on changing the signature `v` to `value` until 
https://github.com/pandas-dev/pandas/pull/24350/files/ff8bbc361b5bed4c244ba38a4a49c0e2e79758e5#diff-99633b1041c2ea6283aa3d5bcf79e8cd is resolved.

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: I think past-me included this in the hopes of making rebasing \epsilon simpler.

**Comment by jbrockmendel** [comment]
updated per comments

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -468,6 +543,67 @@ def _values_for_factorize(self):
...
 
+    def _values_for_argsort(self):
+        return self._data
+
+    # ------------------------------------------------------------------
+    # Additional array methods
+    #  These are not part of the EA API, but we implement them because
+    #  pandas assumes they're there.
+
+    def searchsorted(self, v, side='left', sorter=None):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: Just to be clear, the name for the positional argument hasn't been finalized yet. We figuring out the relative value of matching NumPy vs. matching the rest of pandas here.

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: well, as I have said multiple times, we are already using ``value`` in pandas, so -1 unless its that.

File: pandas/core/indexes/datetimelike.py [hunk 5]
```diff
@@ -478,6 +478,7 @@ def repeat(self, repeats, axis=None):
         nv.validate_repeat(tuple(), dict(axis=axis))
         freq = self.freq if is_period_dtype(self) else None
         return self._shallow_copy(self.asi8.repeat(repeats), freq=freq)
+        # TODO: dispatch to _eadata
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimelike.py]: What's blocking this right now?

File: pandas/core/arrays/datetimelike.py [hunk 6]
```diff
@@ -80,6 +81,80 @@ def _get_attributes_dict(self):
...
+        Examples
+        --------
+        >>> self._unbox_scalar(Timedelta('10s'))  # DOCTEST: +SKIP
+        10000000000
+        """
+        raise AbstractMethodError(self)
+
+    def _check_compatible_with(self, other):
+        # type: (Union[Period, Timestamp, Timedelta, NaTType]) -> None
+        # TODO: Scalar, array, or both?
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: This TODO can be removed I think. We've finalized the type IMO.

But that can be done on merge if we don't have any other commits to push.

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/arrays/datetimelike.py]: Apparently Jeff has decided :)

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Will revert

File: pandas/core/indexes/datetimelike.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: I think nothing, just forgot to change before pushing.  Will do that now.

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: Actually no, this needs to wait until DTI._data is a DTA

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
merging this and closing #24350 ?

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/pull/24350 implements it for ExtensionArray. This does it for datetimelike, so both will be needed I think.

**Comment by jreback** [comment]
okie dokie; this needs a merge master

**Comment by jbrockmendel** [comment]
rebased

**Comment by jreback** [comment]
merged #24350 but looks like this is clean. ping on green.

**Review by TomAugspurger — APPROVED** [review]

File: pandas/tests/arrays/test_datetimelike.py [hunk 7]
```diff
@@ -129,6 +131,57 @@ def test_concat_same_type(self):
...
+        result = arr.searchsorted(arr[1:3])
+        expected = np.array([1, 2], dtype=np.int64)
+        tm.assert_numpy_array_equal(result, expected)
+
+        result = arr.searchsorted(arr[1:3], side="right")
+        expected = np.array([2, 3], dtype=np.int64)
+        tm.assert_numpy_array_equal(result, expected)
+
+        # Following numpy convention, NaT goes at the beginning
+        #  (unlike NaN which goes at the end)
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/tests/arrays/test_datetimelike.py]: https://github.com/numpy/numpy/issues/12629

**Comment by jbrockmendel** [comment]
Ping

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**